### PR TITLE
Last processed position should not be less than the snapshot position 

### LINKED ISSUE #7686: Last processed position in replay machine should not be less than the latest snapshot position
**Describe the bug**

It looks like the lastProcessedPosition is updated regardless if the event is applied or not. 
https://github.com/camunda-cloud/zeebe/blob/d16d730f98499a5dd87600a06af195ee1a8c2ad1/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java#L278

This can cause the `lastProcessedPosition` to be lower than the `snapshotPosition` when the replay state machine was not able to seek directly to the snapshotPosition (see #7684).  This can lead to incorrect state when a new snapshot is taken with the new lastProcessedPosition.

### PR DESCRIPTION
**Pull request by saig0**
## Description

* the last processed position must be greater than the snapshot position
  * a follower may receive events that are already applied on the snapshot
  * the follower should ignore these events and should not update the last processed position
* introduce a new listener for replay to write a proper test case
  * bundle the existing listeners for processing into a new interface
  * get notified when events are replayed
  * the new interface should improve the maintainability  
* reducing the execution time of the test
  * previously, the test waited until the processing starts (i.e. until recovered is called)
  * but in replay mode, the stream processor doesn't switch to processing

## Related issues

closes #7686 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Looks good. Please see my comment on the test.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorReplayModeTest.java [hunk 1]
```diff
@@ -290,6 +290,40 @@ public void shouldSetLastProcessedPositionOnStateToSourcePosition() {
...
+    replayContinuously.snapshot();
+    replayContinuously.closeStreamProcessor();
+
+    // when
+    startStreamProcessor(replayContinuously);
+
+    replayContinuously.writeEvent(
+        ELEMENT_ACTIVATING,
+        RECORD,
+        writer -> writer.sourceRecordPosition(commandPositionBeforeSnapshot));
```
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorReplayModeTest.java]: Yes, the test is written with this in mind. 

Please have a look at the changes here: https://github.com/camunda-cloud/zeebe/pull/7742/files#diff-cad09dfd8f67be3edaa16bc58024157e74c08048000bc49873430bdf9fc7b938R314-R321

[the remark this answers is not in the document]
Comment (deepthidevaki) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorReplayModeTest.java]: Aah.. Thank. Why did I miss it!

**Review by deepthidevaki — APPROVED** [review]
:rocket:

**Pull request merged**